File: Data Structures & Algorithms/two-integer-sum/submission-3.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number[]}
     */
    twoSum(nums, target) {
        const numMap = {};
        for (let i = 0; i < nums.length; i++) {
            const num = nums[i];
            numMap[num] = i;
        }

        for (let i = 0; i < nums.length; i++) {
            const result = target - nums[i];
            if(numMap[result] !== undefined && numMap[result] !== i ) return [i, numMap[result]];
        }
        return [];
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int length = nums.size();
        vector<int> result(length, 1);

        for (int i = 1; i < length; i++) {
            result[i] = result[i - 1] * nums[i - 1];
        }

        int postfix = 1;
        
        for (int i = length - 1; i >= 0; i--) {
            result[i] *= postfix;
            postfix *= nums[i];
        }
        return result;
    }
};
